cmd/dagger/do.go

```diff
@@ -190,6 +200,57 @@ func addCmd(ctx context.Context, cmdStack []*cobra.Command, projCmd dagger.Proje
 					newSelection := &ast.Field{}
 					curSelection.SelectionSet = ast.SelectionSet{newSelection}
 					curSelection = newSelection
+				} else {
+					if outputPath == "" && returnTypeCanUseOutputFlag(projResultType) {
+						return fmt.Errorf("output path not set, --output must be explicitly provided for git:// projects that return files or directories")
+					}
+					outputPath, err = filepath.Abs(outputPath)
+					if err != nil {
+						return fmt.Errorf("failed to get absolute path of output path: %w", err)
+					}
+					switch projResultType {
```

Comment:
> I guess the only catch will be when folks want commands to output to e.g. dist/ or out/. They can use --output obviously, but that'd probably quickly be wrapped with a Bash script. Or you can have your command return an "overlay" like ./dist/foo. Or maybe we can support explicit exports someday?

Yeah that's true, I won't be incredibly saddened if users have to make use of the `--output` flag for those sorts of use cases, but we could definitely try to make it less painful. Maybe a `DAGGER_DEFAULT_OUTPUT` env would help sometimes? Though unless you set that in your `.bashrc` or equivalent it's probably more overhead than it's worth.

We definitely *can* support explicit exports from commands someday, that's what the older iteration of this PR did (if you look at the first edit of the PR description you can see the details). I backed off because
1. It is quite tricky to get a comprehensible API for differentiating between which "host" you want to export (the command's host? the caller's host? etc.). It's also tricky to do this in a backwards compatible way.
2. The implementation is a mess since it involves interaction across sessions. The idea to move the session to a frontend would fix this, but obviously too much to bite off for the moment.

That option is still open for us though, it could easily co-exist with the approach implemented in this PR if desired.

Reply:
> I guess the only catch will be when folks want commands to output to e.g. dist/ or out/. They can use --output obviously, but that'd probably quickly be wrapped with a Bash script. Or you can have your command return an "overlay" like ./dist/foo. Or maybe we can support explicit exports someday?

You can frame this problem as *who* and *where*:

* Q. *who* is responsible for the logic that modifies the project directory?
* A. The author of the command. 

* Q. *where* should that logic live?
* A. In the command code and nowhere else

* Q. I'm using a command, and want to change the logic (for example I want it to write to a `./out` directory)
* A. Change the code of the command, or write a different command and use that

For the above to work, we want to avoid "pseudo-code" living outside the command code (for example a wrapper bash script). And for *that* to happen, we need a rigid convention around the command. In this case: "a command can modify the project directory any way it wants."

Reply:
> In this case: "a command can modify the project directory any way it wants."

Yep, and that's the behavior now if you don't specify `--output`, the file/dir just gets overlayed onto your project dir. `--output` is really only needed for special cases. Good to have but the default of the command writing to the project dir is probably what users will do the vast majority of the time.

Also, I just realized that I misread @vito's original comment, I thought you were talking about the case where you want to write outside the project dir, but I think you're referring to writing to dirs under it.

If so, yeah if the command has a file but wants to ensure it's always written explicitly under `./out`, then it can do this:
```go
return ctx.Client().Directory().WithFile("out/foo.txt", theFile)
```

Which I think is fine. We should optimize `WithFile` to use mergeop and maybe even consider some extra sugar for this pattern, but those would all be improvements to the client SDKs.

Reply:
Agreed 👍